Fabric.Authorization.Domain/Services/GroupMigratorService.cs

```diff
@@ -0,0 +1,156 @@
+        private readonly ILogger _logger;
+
+        public GroupMigratorService(
+            IGroupStore groupStore,
+            ILogger logger)
+        {
+            _groupStore = groupStore ?? throw new ArgumentNullException(nameof(groupStore));
+            _logger = logger;
+        }
+
+        public async Task<GroupMigrationResult> MigrateDuplicateGroups()
+        {
+            var groups = (await _groupStore.GetAll()).ToList();
+
+            var groupKeys = groups
+                .GroupBy(g => g.Name, StringComparer.OrdinalIgnoreCase)
+                .Where(g => g.Count() > 1)
+                .Select(g => g.Key);
+
+            var groupMigrationResult = new GroupMigrationResult();
+            foreach (var key in groupKeys)
+            {
+                var duplicateGroups = groups.Where(g => string.Equals(key, g.Name, StringComparison.OrdinalIgnoreCase)).ToList();
+                var originalGroup = duplicateGroups.First();
+                duplicateGroups.RemoveAt(0);
+
+                var groupMigrationRecord = new GroupMigrationRecord
+                {
+                    DuplicateCount = duplicateGroups.Count,
+                    GroupName = originalGroup.Name
+                };
+
+                foreach (var duplicateGroup in duplicateGroups)
+                {
+                    // migrate roles
+                    foreach (var role in duplicateGroup.Roles)
+                    {
+                        var roleExists = originalGroup.Roles.Any(r => r.Equals(role));
+                        if (!roleExists)
+                        {
+                            _logger.Information($"Migrating Role {role} from {duplicateGroup} to group {originalGroup}");
+                            originalGroup.Roles.Add(role);
+                        }
+                    }
+
+                    // migrate users
+                    foreach (var user in duplicateGroup.Users)
+                    {
+                        var userExists = originalGroup.Users.Any(u => new UserComparer().Equals(u, user));
+                        if (!userExists)
+                        {
+                            _logger.Information($"Migrating User {user} from {duplicateGroup} to group {originalGroup}");
+                            originalGroup.Users.Add(user);
+                        }
+                    }
+
+                    // migrate directory groups to new parents (this will not execute if the duplicate group is a custom group)
+                    foreach (var parent in duplicateGroup.Parents)
+                    {
+                        var parentExists = originalGroup.Parents.Any(p => p.Id == parent.Id);
+                        if (!parentExists)
+                        {
+                            _logger.Information($"Migrating Group {originalGroup} to Parent Group {parent}");
+                            originalGroup.Parents.Add(parent);
+                            /*
```

Comment:
Should we get rid of this commented code?

Reply:
Yes, I was just waiting to make sure the AutoMapper profile stuff was going to work out.